src/backend/optimizer/path/allpaths.c

```diff
@@ -1984,6 +1984,17 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 
 	Assert(IsA(cte->ctequery, Query));
 
+	/*
+	 * We can't do DML for each subplan. It'll cause duplicated DML operations
+	 * or mutation errors during apply_motion() of cdbparallelize().
+	 *
+	 * TODO: Better workaround using materialization or something else if
+	 * possible.
+	 */
+	if (cte->cterefcount > 1 &&
+		((Query *) cte->ctequery)->commandType != CMD_SELECT)
+		elog(ERROR, "too much refs to non-SELECT CTE");
```

Comment:
I think it's quite rough condition. As far I understand, the problem manifests itself only in case of inlining CTEs (see condition below).

Reply:
Have moved this if-clause, what allows to generate plan when ` gp_cte_sharing` is on.

Reply:
It was decided to separate these changes to a separate patch.